### ISSUE TITLE
sysdata's `memory` doesn't account for tmpfs

### ISSUE THREAD

**Issue by dconathan** [ultrabug/py3status#666]
The [memory](https://github.com/ultrabug/py3status/blob/master/py3status/modules/sysdata.py#L85-L128) function of the `sysdata` module does not account for tmpfs directories.  It seems like maybe `memi["Shmem:"]` should also be subtracted... or better yet, why isn't it just `used_mem_kib = total_mem_kib - memi["MemFree:"]` ?

Here's a good breakdown of what all the numbers mean, including which ones refer to tmpfs space:
https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Deployment_Guide/s2-proc-meminfo.html

**Comment by tobes** [comment]
what counts as used memory is never really clear.

Some would consider any memory used, others might say that cached data doesn't count as it can be freed without problem.  Buffers are also temporary.

Personally I'd say that memory used by tempfs is being used and shouldn't be deducted.

on my system the module output is very similar to that given by htop and so looks good.

perhaps you are getting very different results due to your setup.  If so could you give some numbers to help understand the issue.

Alternatively if you want an additional placeholder to `{mem_used}` eg `{mem_used_total}` (we'd need a better name) then we could look at adding that

**Comment by dconathan** [comment]
Okay cool.  Figured some sort of this would be a philosophical "what is used memory" thing.  In my case, there is a mismatch between htop and sysdata.  With a large directory (~7.6G) in my /tmp directory, htop is correct in saying that I'm using about ~7.7G memory.  But sysdata says I'm only using .5 G.  Here's the out put of /proc/meminfo:

```
MemTotal:       16308324 kB
MemFree:          168144 kB
MemAvailable:    7885852 kB
Buffers:           26364 kB
Cached:         15581272 kB
SwapCached:            0 kB
Active:           493808 kB
Inactive:       15400912 kB
Active(anon):     313012 kB
Inactive(anon):  7641876 kB
Active(file):     180796 kB
Inactive(file):  7759036 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                64 kB
Writeback:             0 kB
AnonPages:        282628 kB
Mapped:           242240 kB
Shmem:           7667768 kB
Slab:              91636 kB
SReclaimable:      67204 kB
SUnreclaim:        24432 kB
KernelStack:        3136 kB
PageTables:        10108 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     8154160 kB
Committed_AS:    8706232 kB
VmallocTotal:   34359738367 kB
VmallocUsed:           0 kB
VmallocChunk:          0 kB
HardwareCorrupted:     0 kB
AnonHugePages:     53248 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:      227928 kB
DirectMap2M:    10129408 kB
DirectMap1G:     7340032 kB
```

And when I change the line in sysdata to:
`used_mem_kib = (total_mem_kib - memi["MemAvailable:"])`
It then matches what htop says.

**Comment by lasers** [comment]
Were you looking for this behavior? https://github.com/lasers/py3status/commit/6be2db4de27efdd6c3be9142eff187249f361138

**Comment by dconathan** [comment]
@lasers Yes exactly that behavior.  Thanks!

**Comment by lasers** [comment]
@dconathan Here. https://github.com/lasers/py3status/commit/412bc6572bae0256970c68fb70ab9168e9ec1fda I changed variables because it was misleading. `total_mem_kib` is not same as `total_used_mem_kib`

**Comment by valdur55** [comment]
Why not consume command 'free'  output?

man free output:
`DESCRIPTION
       free  displays  the total amount of free and used physical and swap memory in the system, as well as the buf‐
       fers and caches used by the kernel. The information is gathered by parsing /proc/meminfo. The displayed  col‐
       umns are:

       total  Total installed memory (MemTotal and SwapTotal in /proc/meminfo)

       used   Used memory (calculated as total - free - buffers - cache)

       free   Unused memory (MemFree and SwapFree in /proc/meminfo)

       shared Memory  used (mostly) by tmpfs (Shmem in /proc/meminfo, available on kernels 2.6.32, displayed as zero
              if not available)

       buffers
              Memory used by kernel buffers (Buffers in /proc/meminfo)

       cache  Memory used by the page cache and slabs (Cached and Slab in /proc/meminfo)

       buff/cache
              Sum of buffers and cache

       available
              Estimation of how much memory is available for starting new applications, without swapping. Unlike the
              data  provided by the cache or free fields, this field takes into account page cache and also that not
              all reclaimable memory slabs will be reclaimed due to items being in use (MemAvailable  in  /proc/meminfo, available on kernels 3.14, emulated on kernels 2.6.27+, otherwise the same as free)
`

Looks like `total - available = used` .
   = used  memory.

**Comment by tobes** [comment]
> Why not consume command 'free' output?

from the man page `The information is gathered by parsing /proc/meminfo.`

**Comment by lasers** [comment]
@guiniol Thoughts on this issue? And do you want to try this issue?

**Comment by guiniol** [comment]
I'm pretty sure `free` is taking the info from `/proc/meminfo`.
I don't know how tmpfs are accounted, and where to look to check their usage. The first thing to know is what happens if the memory is full. Will processes get priority or will tmpfs get priority ? If neither get priority (which is what I would think is the case), then it would be nice to show how much tmpfs are using as this is memory that cannot be used by programs (whereas the memory used by the fs cache can be used by programs). However, I would had either an option or a different variable.

As for me taking this. It's an interesting question, but as shown by the (lack of) progress of #1022, I don't have much time right now, so if there's any urgency, better someone else take it.

**Comment by lasers** [comment]
Closed via #1493.
